From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: nopjmp <kthompson@hey.com>
Date: Tue, 25 May 2021 21:48:42 -0500
Subject: [PATCH] Map Decoration Limit

The root cause is the map icon not being removed from the decorations hash map after the item frame is removed.

diff --git a/src/main/java/dev/pomf/dionysus/DionysusConfig.java b/src/main/java/dev/pomf/dionysus/DionysusConfig.java
index 0fe33eb470ad0be018d127194273b39cb42be0ff..b9b330b29b529b6390f8e69b7acd0fce9121bfbf 100644
--- a/src/main/java/dev/pomf/dionysus/DionysusConfig.java
+++ b/src/main/java/dev/pomf/dionysus/DionysusConfig.java
@@ -251,4 +251,9 @@ public class DionysusConfig {
             Bukkit.getLogger().info("Tickless Block Placement Mechanics activated. Block placement based on time instead of ticks.");
         }
     }
+
+    public static int mapDecorationsLimit = 20;
+    private static void mapDecorationsLimit() {
+        mapDecorationsLimit = getInt("map-decoration-limit", mapDecorationsLimit);
+    }
 }
diff --git a/src/main/java/net/minecraft/server/WorldMap.java b/src/main/java/net/minecraft/server/WorldMap.java
index 13c06e94ed5e9afd23afd4336dc5280e353beb0d..f5de64f2729c72a58188570565d17d9dac5eaa4a 100644
--- a/src/main/java/net/minecraft/server/WorldMap.java
+++ b/src/main/java/net/minecraft/server/WorldMap.java
@@ -16,6 +16,8 @@ import org.bukkit.craftbukkit.CraftWorld;
 import org.bukkit.craftbukkit.map.CraftMapView;
 // CraftBukkit end
 
+import dev.pomf.dionysus.DionysusConfig; // Dionysus
+
 public class WorldMap extends PersistentBase {
 
     public int centerX;
@@ -282,6 +284,11 @@ public class WorldMap extends PersistentBase {
             }
         }
 
+        // Dionysus start - limit decorations
+        if (this.decorations.size() >= DionysusConfig.mapDecorationsLimit) {
+            return;
+        }
+        // Dionysus end
         this.decorations.put(s, new MapIcon(mapicon_type, b0, b1, b2));
     }
 
